java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java

```diff
@@ -115,49 +129,36 @@ public void computedFrom(List<SymbolicValue> symbolicValues) {
   }
 
   public List<ProgramState> setConstraint(ProgramState programState, ObjectConstraint nullConstraint) {
-    Object data = programState.getConstraint(this);
-    if (data instanceof ObjectConstraint) {
-      ObjectConstraint nc = (ObjectConstraint) data;
-      if (nc.isNull() ^ nullConstraint.isNull()) {
-        // setting null where value is known to be non null or the contrary
-        return ImmutableList.of();
-      } else if (hasAnyStatus(nc) && hasNoStatus(nullConstraint)) {
-        return ImmutableList.of(programState);
-      }
-    }
-    if (data instanceof BooleanConstraint) {
-      return nullConstraint.isNull() ? ImmutableList.of() : ImmutableList.of(programState);
-    }
-    if (data == null || !data.equals(nullConstraint)) {
+    Constraint cstraint = programState.getConstraint(this, nullConstraint.getClass());
+    if (cstraint == null) {
       return ImmutableList.of(programState.addConstraint(this, nullConstraint));
+    } else if (cstraint != nullConstraint) {
+      return ImmutableList.of();
     }
     return ImmutableList.of(programState);
   }
 
-  private static boolean hasAnyStatus(ObjectConstraint constraint) {
-    return !hasNoStatus(constraint);
-  }
-
-  private static boolean hasNoStatus(ObjectConstraint constraint) {
-    return constraint.hasStatus(null);
-  }
-
   public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
-    Object data = programState.getConstraint(this);
-    // update program state only for a different constraint
-    if (data instanceof BooleanConstraint && !data.equals(booleanConstraint)) {
-      // setting null where value is known to be non null or the contrary
-      return ImmutableList.of();
-    }
-    if ((data == null || isNonNullConstraint(data)) && programState.canReach(this)) {
-      // store constraint only if symbolic value can be reached by a symbol.
+    Constraint cstraint = programState.getConstraint(this, booleanConstraint.getClass());
+    if (cstraint == null) {
       return ImmutableList.of(programState.addConstraint(this, booleanConstraint));
+    } else if (cstraint != booleanConstraint) {
+      return ImmutableList.of();
     }
     return ImmutableList.of(programState);
   }
 
-  private static boolean isNonNullConstraint(Object data) {
-    return data instanceof ObjectConstraint && !((ObjectConstraint) data).isNull();
+  public List<ProgramState> setConstraint(ProgramState programState, Constraint constraint) {
+    if(constraint instanceof BooleanConstraint) {
+      return setConstraint(programState, (BooleanConstraint) constraint);
+    } else if(constraint instanceof ObjectConstraint) {
+      return setConstraint(programState, (ObjectConstraint) constraint);
+    }
+    Constraint cstraint = programState.getConstraint(this, constraint.getClass());
+    if (cstraint == null || cstraint != constraint) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)